core/src/Kokkos_ExecPolicy.hpp

```diff
@@ -847,6 +847,301 @@ KOKKOS_INLINE_FUNCTION_DELETED Impl::TeamThreadRangeBoundariesStruct<
+  static constexpr bool isDirectionLeft =
+      (Rank::outer_direction == Iterate::Left);
+  // If vector parallelism is in use deploy thread parallelism on
+  // the second to the last rank, otherwise, thread parallelism on the last rank
+  static constexpr int par_rt =
+      isDirectionLeft ? ((t_and_v) ? 1 : 0)
+                      : ((t_and_v) ? Rank::rank - 2 : Rank::rank - 1);
+  // Vector parallelism will always be on the last index
+  static constexpr int par_rv  = isDirectionLeft ? 0 : Rank::rank - 1;
+  static constexpr int invalid = -2;
+};
+#endif
+
+#ifdef KOKKOS_ENABLE_HIP
+template <typename Rank, MDTeamRangeThreadAndVector t_and_v>
+struct ParallelRank<Rank, Experimental::HIP, t_and_v> {
+  static constexpr bool isDirectionLeft =
+      (Rank::outer_direction == Iterate::Left);
+  // If vector parallelism is in use deploy thread parallelism on
+  // the second to the last rank, otherwise, thread parallelism on the last rank
+  static constexpr int par_rt =
+      isDirectionLeft ? ((t_and_v) ? 1 : 0)
+                      : ((t_and_v) ? Rank::rank - 2 : Rank::rank - 1);
+  // Vector parallelism will always be on the last index
+  static constexpr int par_rv  = isDirectionLeft ? 0 : Rank::rank - 1;
+  static constexpr int invalid = -2;
+};
+#endif
+
+// TODO check par_rt and par_rv validity
+#ifdef KOKKOS_ENABLE_OPENMPTARGET
+template <typename Rank, MDTeamRangeThreadAndVector t_and_v>
+struct ParallelRank<Rank, Experimental::OpenMPTarget, t_and_v> {
+  static constexpr bool isDirectionLeft =
+      (Rank::outer_direction == Iterate::Left);
+  static constexpr int par_rt  = isDirectionLeft ? Rank::rank - 1 : 0;
+  static constexpr int par_rv  = isDirectionLeft ? 0 : Rank::rank - 1;
+  static constexpr int invalid = -2;
+};
+#endif
+
+#ifdef KOKKOS_ENABLE_OPENMP
+template <typename Rank, MDTeamRangeThreadAndVector t_and_v>
+struct ParallelRank<Rank, OpenMP, t_and_v> {
+  static constexpr bool isDirectionLeft =
+      (Rank::outer_direction == Iterate::Left);
+  static constexpr int par_rt  = isDirectionLeft ? Rank::rank - 1 : 0;
+  static constexpr int par_rv  = isDirectionLeft ? 0 : Rank::rank - 1;
+  static constexpr int invalid = -2;
+};
+#endif
+
+#ifdef KOKKOS_ENABLE_THREADS
+template <typename Rank, MDTeamRangeThreadAndVector t_and_v>
+struct ParallelRank<Rank, Threads, t_and_v> {
+  static constexpr bool isDirectionLeft =
+      (Rank::outer_direction == Iterate::Left);
+  static constexpr int par_rt  = isDirectionLeft ? Rank::rank - 1 : 0;
+  static constexpr int par_rv  = isDirectionLeft ? 0 : Rank::rank - 1;
+  static constexpr int invalid = -2;
+};
+#endif
+
+// TODO check par_rt and par_rv validity
+#ifdef KOKKOS_ENABLE_SYCL
+struct ParallelRank<Rank, SYCL, t_and_v> {
+  static constexpr bool isDirectionLeft =
+      (Rank::outer_direction == Iterate::Left);
+  static constexpr int par_rt  = isDirectionLeft ? Rank::rank - 1 : 0;
+  static constexpr int par_rv  = isDirectionLeft ? 0 : Rank::rank - 1;
```

Comment:
@masterleinad should this be like CUDA? i.e. do we support separate Thread and Vector parallelism?